cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/benchmark/Benchmark.java

```diff
@@ -0,0 +1,199 @@
+            start = Instant.now();
+            earnRequest = incentiveSystem.generateEarnRequest(
+                    token,
+                    ppk,
+                    userKeyPair
+            );
+            finish = Instant.now();
+            tEarnRequest[i] = Duration.between(start, finish).toNanos();
+            start = Instant.now();
+            earnResponse = incentiveSystem.generateEarnRequestResponse(
+                    earnRequest,
+                    EARN_SPEND_AMOUNT,
+                    providerKeyPair
+            );
+            finish = Instant.now();
+            tEarnResponse[i] = Duration.between(start, finish).toNanos();
+            start = Instant.now();
+            token = incentiveSystem.handleEarnRequestResponse(
+                    earnRequest,
+                    earnResponse,
+                    EARN_SPEND_AMOUNT,
+                    token,
+                    ppk,
+                    userKeyPair
+            );
+            finish = Instant.now();
+            tEarnHandleResponse[i] = Duration.between(start, finish).toNanos();
+        }
+
+        for (int i = 0; i < benchmarkConfig.iterations; i++) {
+            feedbackFunction.accept(BenchmarkState.SPEND_DEDUCT, i);
+            var tid = incentiveSystem.pp.getBg().getZn().getUniformlyRandomElement();
+            start = Instant.now();
+            assert token != null;
+            spendRequest = incentiveSystem.generateSpendRequest(
+                    token,
+                    ppk,
+                    EARN_SPEND_AMOUNT,
+                    userKeyPair,
+                    tid
+            );
+
+            finish = Instant.now();
+            tSpendRequest[i] = Duration.between(start, finish).toNanos();
+            start = Instant.now();
+            spendResponseTuple = incentiveSystem.generateSpendRequestResponse(
+                    spendRequest,
+                    providerKeyPair,
+                    EARN_SPEND_AMOUNT,
+                    tid
+            );
+            finish = Instant.now();
+            tSpendResponse[i] = Duration.between(start, finish).toNanos();
+            start = Instant.now();
+            token = incentiveSystem.handleSpendRequestResponse(
+                    spendResponseTuple.getSpendResponse(),
+                    spendRequest,
+                    token,
+                    EARN_SPEND_AMOUNT,
+                    ppk,
+                    userKeyPair
+            );
+            finish = Instant.now();
+            tSpendHandleResponse[i] = Duration.between(start, finish).toNanos();
+        }
+        return new BenchmarkResult(
+                tJoinRequest,
+                tJoinResponse,
+                tJoinHandleResponse,
+                tEarnRequest,
+                tEarnResponse,
+                tEarnHandleResponse,
+                tSpendRequest,
+                tSpendResponse,
+                tSpendHandleResponse
+        );
+    }
+}
```

Comment:
Does not look like too complicated code but some comments might improve readability and ease navigation.

Reply:
Could you specify this? I don't see any code that is not self-explanatory, so I don't want to the obvious things :smile: